### DISCUSSION TITLE
Book — Coding for engineers — first chapters online

### DISCUSSION THREAD

**Discussion by LinKayser** [leap71/PicoGK#24] [Coding for (Computational) Engineers]
![E8F02E5D-4525-45FB-961D-521B4686B525](https://github.com/leap71/PicoGK/assets/148448288/80561d73-12b9-48af-9ce2-9c454cddeecd)

Here's a post I just wrote on [LinkedIn](https://www.linkedin.com/feed/update/urn:li:activity:7184553702899277825/).

Please free to give me feedback here in this thread.

----

OK, I have just committed to teaching engineers how to code — wish me luck! Most engineers don't know how to write software. This is wrong. Coding teaches you computational thinking. Computational thinking is an important way to analyze and understand the world. To an engineer it is as essential as math. Period. But it is even more important for the emerging paradigm of Computational Engineering. As an engineer you should not have to engage in the manual labor of pressing buttons in a CAD program. You should encode your knowledge into computer algorithms, that are reusable and eventually do most of the work that currently consumes your life time. This elevates your skill and allows you create machines that have never been built before.

I wrote PicoGK, our #opensource kernel for Computational Engineering last year. Since then, people have asked me constantly about a good programming course to get started. I couldn't find one that I liked. So, I have embarked on this project, where I will write a book on the topic. By the end of it, you will know how to be a Computational Engineer using PicoGK ("peacock").

Not in a bare-bones simple way — but to really architect Computational Engineering Models, like we create at LEAP 71. This is not "scripting" or hacking together Python scripts. It's crafting object-oriented, engineering-centric, software to elevate the profession of engineering.

I am doing this the way I think it should be done, and it may be a bit unconventional. So, please give me feedback, I promise to incorporate it. I am doing this publicly on the LEAP 71 Github, which feeds into the [PicoGK.org](http://picogk.org/) website.

I am planning to release one chapter a week (I hope I can keep it up).

I just put the first chapters online, check them out: https://picogk.org/coding-for-engineers/

**Comment by RadFad1** [comment]
While reading the Fundamentals chapter, I noticed how each new question I had was answered in the subsequent paragraphs. Many concepts that would typically be glossed over were elaborated on as they were introduced. I like how Lin expressed his opinions on certain conventions e.g. Variable naming styles. I also like that Lin is bringing his years of experience in teaching the subject, so consequently there is a certain confidence and security that comes with learning from him.
I look forward to the coming chapters.
Also, good luck!

**Reply by LinKayser** [reply]
Thanks for the positive feedback, very much appreciated.

**Comment by LinKayser** [comment]
The next chapter is online. https://picogk.org/coding-for-engineers/3-running-code.html

**Comment by valdiskrumins** [comment]
Thank you for your work Lin! I believe in voxels too 🫡
Who knows, this might even be the time I actually learn some real programming...



Just a note.

Unless the location of the zeros are as below on purpose (for learning debugging and troubleshooting) https://picogk.org/coding-for-engineers/4-classes.html

`TeslaRoadster oFredsCar = new(0, "Fred's Tesla");
Console.WriteLine(oFredsCar.strName());

TeslaRoadster oEmmasCar = new(0, "Emma's fun scooter");
Console.WriteLine(oEmmasCar.strName());`


**There might be a typo as the order has to be reversed for the values to be picked up**:
`TeslaRoadster oFredsCar = new("Fred's Tesla", 0);
Console.WriteLine(oFredsCar.strName());

TeslaRoadster oEmmasCar = new("Emma's fun scooter", 0);
Console.WriteLine(oEmmasCar.strName());`

<img width="663" alt="Screenshot 2024-04-29 at 17 09 55" src="https://github.com/leap71/PicoGK/assets/43199029/1ed2117f-83d7-426b-a26c-ccc35585a566">
<img width="649" alt="Screenshot 2024-04-29 at 17 10 05" src="https://github.com/leap71/PicoGK/assets/43199029/3025ea90-29ea-41c6-b59a-59764cd090df">

**Reply by LinKayser** [reply]
Very true. Thanks for the feedback! I just fixed it.
Lin

**Comment by exzactlines** [comment]
Hi All,
First off, thank you @LinKayser for working to synthesize your experience, mastery, and knowledge into a small course for the engineers (like me) with limited coding experience. I am not as familiar with Github as I should be so forgive me if I need to post this in another space - but I am ignorant to convention. 
On to me question: in Chapter 3: Running Code, I did not understand the section "Compiling your code" too well. Several lingering questions remained about Visual Studio Code after reading that section. Where does the "Program.cs" file get created? I am hoping you can expand on this section when you have time - I think a bare minimum screen record of this could be great - at least for me personally. 

![image](https://github.com/leap71/PicoGK/assets/13138035/a25c0859-43b4-4c92-b40d-58b92889ae46)

Thank you again for taking on this task and helping me to learn something new. I am excited to start implementing these lessons and start making geometry soon. 
-Zach

**Reply by valdiskrumins** [reply]
<img width="1512" alt="Screenshot 2024-05-05 at 19 14 55" src="https://github.com/leap71/PicoGK/assets/43199029/8f65615c-6d70-4541-bc8a-431d083ea68b">
<img width="1512" alt="Screenshot 2024-05-05 at 19 15 06" src="https://github.com/leap71/PicoGK/assets/43199029/b210140c-aee4-42f4-a07d-ffbf9f445bb4">
<img width="1512" alt="Screenshot 2024-05-05 at 19 15 15" src="https://github.com/leap71/PicoGK/assets/43199029/14f633d5-7cbf-42cc-9331-645b60ebefd3">
<img width="1512" alt="Screenshot 2024-05-05 at 19 15 21" src="https://github.com/leap71/PicoGK/assets/43199029/73a80171-9bd9-4ee2-8395-2bfdbf644623">
<img width="1512" alt="Screenshot 2024-05-05 at 19 15 45" src="https://github.com/leap71/PicoGK/assets/43199029/5ea791bf-1e51-4394-a94b-432e7606d63c">
<img width="1512" alt="Screenshot 2024-05-05 at 19 15 53" src="https://github.com/leap71/PicoGK/assets/43199029/567dec88-d997-4529-98d8-4b1f59c6a3ec">
<img width="1512" alt="Screenshot 2024-05-05 at 19 15 58" src="https://github.com/leap71/PicoGK/assets/43199029/9c76f8c5-fe46-4070-8de8-092e082ae723">
<img width="1512" alt="Screenshot 2024-05-05 at 19 16 07" src="https://github.com/leap71/PicoGK/assets/43199029/08af3749-3f30-4e1e-9ca4-22902771ad63">
<img width="1512" alt="Screenshot 2024-05-05 at 19 17 00" src="https://github.com/leap71/PicoGK/assets/43199029/859535fe-2600-4b00-9ef3-4282baec69cb">
<img width="1512" alt="Screenshot 2024-05-05 at 19 17 07" src="https://github.com/leap71/PicoGK/assets/43199029/f7910575-6b30-44df-9fbd-e1a9eb12ed0d">
<img width="1512" alt="Screenshot 2024-05-05 at 19 17 12" src="https://github.com/leap71/PicoGK/assets/43199029/df7d7cac-7abd-48c7-a6f1-81377ec13580">
<img width="1512" alt="Screenshot 2024-05-05 at 19 17 16" src="https://github.com/leap71/PicoGK/assets/43199029/96b55e9c-6bd4-4784-a525-4c8e7e9d9518">
<img width="1512" alt="Screenshot 2024-05-05 at 19 17 20" src="https://github.com/leap71/PicoGK/assets/43199029/b6979341-c421-43bc-a23b-02118525cd40">

**Reply by LinKayser** [reply]
Thank you for this. You are right, I am skipping over this maybe a bit to hastily. I went back and forth in my thinking on this. Unfortunately Microsoft changes things so often, that I would constantly have to update everything. But I do this already in the PicoGK documentation. Maybe as a compromise, I will simply link to the PicoGK guide on how to create a console application.

The book tries to stay away from "tutorial-like", click here, click there instructions. But this is the one place, where it's necessary, because it is so damn hard to create a project.

https://github.com/leap71/PicoGK/blob/main/Documentation/VisualStudioCode_FirstTime.md

Thanks for the excellent feedback.
Lin

**Comment by valdiskrumins** [comment]
Lesson 5

Thank you! I am hoping once these concepts are understood the basic things will be easier :) And I appreciate this approach is based on the decades of best practices etc.

It took me some time to realise that there will be no visual/console feedback from the Charge(); unless I add one myself.
<img width="679" alt="image" src="https://github.com/leap71/PicoGK/assets/43199029/43333392-84dc-48a1-b4d4-78ce87925401">

I appreciate that this might not be the point of this lesson, but is there a way to reload the code so that the 'health counter' does not restart/keeps counting? I did change the formula so it would run down faster.

How do I call the base type of Car.PrepareForDriving(); if needed?
<img width="313" alt="image" src="https://github.com/leap71/PicoGK/assets/43199029/3605546e-56d9-43f0-918d-10e6a9156198">
Unless the whole idea here was that it would never be called, instead brothers and sister of TeslaRoadster would be.

**Reply by LinKayser** [reply]
Your car will have a broken battery very soon at this rate :)

You will never directly call a base class implementation from the "outside" of a class. The whole idea of information hiding is that you don't have to have any knowledge about the internal function of a class.

There is, however, of course the very valid point that you may want to call the base class implementation from your derived class implementation. So, in other words, you want to call the `Car`'s `PrepareForDriving` function from `TeslaRoadster.PrepareForDriving`. You can do this by invoking `base.PrepareForDriving`. This only works from within the same function in a derived class, though. 

In practice, you will rarely do this, as there are more fail proof ways to deal with this stuff. But that's another chapter in the book (I just decided).

Lin

**Comment by mosri** [comment]
Hello @LinKayser, thank you for all your efforts to educate the world on this paradigm shift of engineering.

I have a question; I have been trying to understand what exactly are computational engineering models beyond the abstract notions that are so frequently mentioned in your articles and such. Are there plans to go over how to build our own CEMs in your coding for engineers series and perhaps releasing with it some solid example of a CEM?


Also, as far as feedback. As someone who has been programming for over 10 years, I find your series on programming to be very clear and concise. It isn’t dense so it’s easy to understand the ideas and has served as a good review resource for me. 

thank you 

Mark

**Reply by LinKayser** [reply]
Yes, I will slowly build up a CEM over the coming weeks to showcase it.

Fundamentally it starts with the question what is a Computational Model. It's a bunch of algorithms that translates inputs into output, generating something in between.

For example, a Computational Engineering Model of a rocket engine, like Noyron RP, will take the inputs like desired thrust, propellant types, massflow, etc. It then runs it through rocket equations, thermal models, flow calculations, etc.

From these intermediate values, it will then generate geometry, based on physical formulas, but also using engineering logic.

The result is then an analysis about the expected performance of the rocket engine, but also the engine geometry.

More on this topic soon, thanks for asking — we haven't explained that well yet.

Best,
Lin

**Comment by valdiskrumins** [comment]
0s    0.0+ Happy Computational Engineering!
<img width="1290" alt="Screenshot 2024-06-07 at 18 22 46" src="https://github.com/leap71/PicoGK/assets/43199029/71367773-ccfc-4d24-b8cc-566b3a1c7b0f">
<img width="1289" alt="Screenshot 2024-06-07 at 18 23 20" src="https://github.com/leap71/PicoGK/assets/43199029/195a53f1-fe8e-4f42-a019-b5819dc87e75">

**Comment by madscientist42** [comment]
I do wish you luck.  I find it...strange...that people don't code nearly as much as they do.

For me, I came to Engineering from the other direction.  I started by coding...then by learning the math associated with computing (I am a Computer Scientist by degree)...and then reached to Engineering.  For me, coding comes...heh...naturally...as naturally as the Engineering does now.

So...hopefully, you will show many the same journey I'm now on and have been for decades...which now takes a new and interesting turn with PicoGK.

**Reply by valdiskrumins** [reply]
Thanks for the encouragement!

I was following along nicely until the "Let’s build a computational fixture maker" which I could not recreated while following the description/instructions - I saw these lessons have been edited/improved now. It has been extremely busy time since then and I have not had chance to look at it again. Hopefully I shall continue in a couple of weeks.

I used to like assembler although I never built anything meaningful using it. I have actually created somewhat complex things in LUA and it does seem that what stops me is the other languages is the complex looking syntax (brackets, commas... etc.)... I have been creating complex self-generating designs in Creo but again, only because 'Creo Relationships' literally have no syntax - it is just equations with IFs and ENDIFs referencing the items from the existing model/feature tree...

**Reply by LinKayser** [reply]
You should be fine — C# is so much more "readable" than the other languages you mentioned. LUA is a nice language, but to use it with modern concepts (like objects), you kind of have to hack it.
